client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -200,8 +201,14 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        synchronizer.updateState((state, updates) -> {
+            updates.add(new UpdatingConfig(true));
+        });
+        getThrowingException(controller.updateReaderGroup(scope, groupName, config));
         Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        synchronizer.updateState((state, updates) -> {
```

Comment:
the problem really is "after" the call to controller when this code is calling `synchronizer.updateState(RGStateInit)`. 

if RG(b) has already successfully reset, then (a) should not be repeating it, rather exit. 
so instead of blindly resetting here, we should actually check if a reset with our desired generation has already completed or not.. and if not, then we should reset, else exit.

Reply:
Implementing the logic as explained above.